### PR TITLE
pkg/authorization: add audit logging

### LINKED ISSUE #1307: Enhance audit logging, especially for authorization
Currently, we leave little traces about authorizers besides native kube RBAC ones. 

Native kube RBAC authorizer creates audit log entries for two cases:

**allowed:**
- _client_: client gets no reason (expected)
- _audit_: reason is encoded in the audit annotation:
`"authorization.k8s.io/reason": "RBAC: allowed by RoleBinding \"foo\" of Role \"bar\" to User \"baz\""`

**denied:**
- _client_: client gets a simplified reason of denial, example:
`Error from server (Forbidden): pods is forbidden: User "foo" cannot list resource "bar" in API group "" in the namespace "baz"`
- _audit_: reason is only encoded in the message, not in the audit annotation without further details:
`"message": "pods is forbidden: User \"foo\" cannot list resource \"bar\" in API group \"\" in the namespace \"baz\"",
"authorization.k8s.io/reason": "" `

For kcp I suggest to create dedicated annotations per authorizer (api-binding, top level, content) and set potential reasoning messages for both denial and allowed requests, examples:
- `apibinding.authorization.kcp.io/decision: "allow/forbid"`
- `toplevel.authorization.kcp.io/decision: "allow/forbid"` 
- `content.authorization.kcp.io/decision: "allow/forbid"`
- `apibinding.authorization.kcp.io/reason: "allowed, no maximal local permission policy present"`
- `apibinding.authorization.kcp.io/reason: "forbidden by RBAC in API Export cluster \"foo\""` (reason will be encoded via  message , see above)
- `content.authorization.kcp.io/reason: "allowed by cluster local service account (extraGroups=[\"system:kcp:clusterworkspace:access\"])"`
- `content.authorization.kcp.io/reason: "allowed by \"admin\" verb in workspace \"foo\" (extraGroups=[\"system:kcp:clusterworkspace:access\",\"system:kcp:clusterworkspace:admin\"])" `

### PR DESCRIPTION
**Pull request by s-urbaniak**
## Summary

This adds audit logs for remaining authorizers.

## Related issue(s)

Fixes #1307

### CODE REVIEW

File: pkg/authorization/apibinding_authorizer.go [hunk 1]
```diff
@@ -98,33 +105,68 @@ func (a *apiBindingAccessAuthorizer) Authorize(ctx context.Context, attr authori
...
+		)
 		return authorizer.DecisionNoOpinion, apiBindingAccessDenied, err
 	}
 
 	// If we can't find the export default to close
 	if !found {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			APIBindingContentAuditDecision, DecisionNoOpinion,
+			APIBindingContentAuditReason, fmt.Sprintf("API export %q not found", bindingLogicalCluster.Workspace.ExportName),
```
Comment (stevekuznetsov) [inline pkg/authorization/apibinding_authorizer.go:150]: Do we know that `.Workspace` is not `nil`?

Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:150]: that's a very good question, i am not sure but i'll gate this properly 👍

Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:150]: oh wait, in fact we do: https://github.com/kcp-dev/kcp/pull/1833/files/4120653e4864e6ffc04b5c3e297fec13c76bc7b2#diff-6104f7f2f6a565ab7fe344c22770310122299daf5d0f425b2a61329e88ef13a7R240

Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:150]: (but to be on the safe side, this is audit, i'll gate it)

Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:150]: fixed

File: pkg/authorization/apibinding_authorizer.go [hunk 2]
```diff
@@ -98,33 +105,73 @@ func (a *apiBindingAccessAuthorizer) Authorize(ctx context.Context, attr authori
...
+			APIBindingContentAuditReason, fmt.Sprintf("error getting API binding reference: %v", err),
+		)
 		return authorizer.DecisionNoOpinion, apiBindingAccessDenied, err
 	}
 
 	if !bound {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			APIBindingContentAuditDecision, DecisionAllowed,
+			APIBindingContentAuditReason, "no API binding found",
```
Comment (stevekuznetsov) [inline pkg/authorization/apibinding_authorizer.go:130]:
```suggestion
			APIBindingContentAuditReason, "no API binding bound",
```

?

File: pkg/authorization/apibinding_authorizer.go [hunk 3]
```diff
@@ -98,33 +105,73 @@ func (a *apiBindingAccessAuthorizer) Authorize(ctx context.Context, attr authori
...
+	if bindingLogicalCluster.Workspace != nil {
+		exportName = bindingLogicalCluster.Workspace.ExportName
+	}
+
 	// If we can't find the export default to close
 	if !found {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			APIBindingContentAuditDecision, DecisionNoOpinion,
+			APIBindingContentAuditReason, fmt.Sprintf("API export %q not found", exportName),
```
Comment (stevekuznetsov) [inline pkg/authorization/apibinding_authorizer.go:155]: Do we need path?

File: pkg/authorization/apibinding_authorizer.go [hunk 4]
```diff
@@ -98,33 +105,73 @@ func (a *apiBindingAccessAuthorizer) Authorize(ctx context.Context, attr authori
...
+		)
 		return authorizer.DecisionNoOpinion, apiBindingAccessDenied, err
 	}
 
 	if apiExport.Spec.MaximalPermissionPolicy == nil {
+		kaudit.AddAuditAnnotations(
+			ctx,
+			APIBindingContentAuditDecision, DecisionAllowed,
+			APIBindingContentAuditReason, fmt.Sprintf("no maximal permission policy present in API export %q", apiExport.Name),
```
Comment (stevekuznetsov) [inline pkg/authorization/apibinding_authorizer.go:164]: APIExport path/owning lcluster?

Comment (stevekuznetsov) [reply pkg/authorization/apibinding_authorizer.go:164]: (ditto for rest)

File: pkg/authorization/apibinding_authorizer.go (hunk 3 above)
Comment (ncdc) [reply pkg/authorization/apibinding_authorizer.go:155]: +1

File: pkg/authorization/decision.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorization
+
+import "k8s.io/apiserver/pkg/authorization/authorizer"
+
+const (
+	DecisionNoOpinion = "noopinion"
```
Comment (ncdc) [inline pkg/authorization/decision.go:22]: Can we revert these back to NoOpinion, Allowed, Denied?

File: pkg/authorization/local_authorizer.go [hunk 6]
```diff
@@ -80,5 +93,14 @@ func (a *LocalAuthorizer) Authorize(ctx context.Context, attr authorizer.Attribu
 		&rbac.ClusterRoleBindingLister{Lister: filteredInformer.ClusterRoleBindings().Lister()},
 	)
 
-	return scopedAuth.Authorize(ctx, attr)
+	dec, reason, err := scopedAuth.Authorize(ctx, attr)
+
+	decString := decisionString(dec)
```
Comment (ncdc) [inline pkg/authorization/local_authorizer.go:98]: global comment: if you're not reusing this, you can omit creating a variable

File: pkg/authorization/decision.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/authorization/decision.go:22]: sure but this was to make it somewhat aligned to upstream reasons in audit. I have no strong opinions here.

File: pkg/authorization/local_authorizer.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/authorization/local_authorizer.go:98]: hrm, yes, this is a left-over from a previous code state, this var doesn't make sense any more.

File: pkg/authorization/apibinding_authorizer.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:130]: fixed 👍

File: pkg/authorization/apibinding_authorizer.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:155]: done 👍

File: pkg/authorization/local_authorizer.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/authorization/local_authorizer.go:98]: fixed

File: pkg/authorization/apibinding_authorizer.go (hunk 4 above)
Comment (s-urbaniak) [reply pkg/authorization/apibinding_authorizer.go:164]: done

**Pull request merged**